pkg/scheduler/predicates/ha.go

```diff
@@ -15,117 +15,111 @@ package predicates
 func (h *ha) Filter(instanceName string, pod *apiv1.Pod, nodes []apiv1.Node) ([]apiv1.Node, error) {
 	ns := pod.GetNamespace()
 	podName := pod.GetName()
+	component := pod.Labels[label.ComponentLabelKey]
+	tcName := getTCNameFromPod(pod, component)
 
-	var component string
-	var exist bool
-	if component, exist = pod.Labels[label.ComponentLabelKey]; !exist {
-		return nodes, fmt.Errorf("can't find component in pod labels: %s/%s", ns, podName)
+	if len(nodes) == 1 {
+		pvcName := fmt.Sprintf("%s-%s", component, podName)
+		pvc, err := h.pvcGetFn(ns, pvcName)
+		if err != nil {
+			return nil, err
+		}
+		if pvc.Status.Phase == apiv1.ClaimBound {
+			return nodes, nil
+		}
 	}
+
 	podList, err := h.podListFn(ns, instanceName, component)
 	if err != nil {
 		return nil, err
 	}
-
-	ordinal, err := util.GetOrdinalFromPodName(podName)
+	tc, err := h.tcGetFn(ns, tcName)
 	if err != nil {
 		return nil, err
 	}
-
-	// when a deleted pod is recreated again, it should be rescheduled to the original node
-	if len(nodes) == 1 {
-		nextPodName := util.GetNextOrdinalPodName(podName, ordinal)
-		for _, pod := range podList.Items {
-			if pod.GetName() == nextPodName && pod.Spec.NodeName != "" {
-				return nodes, nil
-			}
-		}
-	}
+	replicas := getReplicasFrom(tc, component)
 
 	nodeMap := make(map[string][]string)
 	for _, node := range nodes {
 		nodeMap[node.GetName()] = make([]string, 0)
 	}
 	for _, pod := range podList.Items {
-		podName1 := pod.GetName()
+		pName := pod.GetName()
 		nodeName := pod.Spec.NodeName
-		ordinal1, err := util.GetOrdinalFromPodName(podName1)
-		if err != nil {
-			return nil, err
-		}
-
-		if ordinal1 < ordinal && nodeName == "" {
-			return nil, fmt.Errorf("waiting for pod: %s/%s to be scheduled", ns, podName1)
-		}
 		if nodeName == "" || nodeMap[nodeName] == nil {
 			continue
 		}
 
-		nodeMap[nodeName] = append(nodeMap[nodeName], podName1)
+		nodeMap[nodeName] = append(nodeMap[nodeName], pName)
 	}
 	glog.V(4).Infof("nodeMap: %+v", nodeMap)
 
-	var min int
-	var minInitialized bool
-	for _, podNameArr := range nodeMap {
-		count := len(podNameArr)
-		if !minInitialized {
-			minInitialized = true
-			min = count
+	min := -1
+	minNodeNames := make([]string, 0)
+	for nodeName, podNames := range nodeMap {
+		podsCount := len(podNames)
+		if podsCount+1 >= int(replicas+1)/2 {
+			continue
 		}
-		if count < min {
-			min = count
+		if min == -1 {
+			min = podsCount
 		}
-	}
-	if ordinal < replicas && min != 0 {
-		return nil, fmt.Errorf("the first %d pods can't be scheduled to the same node", replicas)
-	}
 
-	minNodeNames := make([]string, 0)
-	for nodeName, podNameArr := range nodeMap {
-		if len(podNameArr) == min {
-			minNodeNames = append(minNodeNames, nodeName)
+		if podsCount > min {
+			continue
+		}
+		if podsCount < min {
+			min = podsCount
+			minNodeNames = make([]string, 0)
 		}
+		minNodeNames = append(minNodeNames, nodeName)
 	}
 
+	if len(minNodeNames) == 0 {
```

Comment:
If the passed parameter `nodes []apiv1.Node` is empty slice, should we return quickly at the beginning of the function?